### DISCUSSION TITLE
Element Zapper no longer needed? (Feedback/reception desired)

### DISCUSSION THREAD

**Discussion by arjpar** [arjpar/WebShield#5] [General]
I would love to hear your thoughts on the removal of an Element Zapper (not yet implemented) on WebShield, similar to uBO's element zapper. The reason for the removal/non implementation of the element zapper feature is that Safari (latest Sequoia) has a built in element zapper, and I would rather default to Safari instead of re-implementing something that is already provided by default. **Thoughts?**

It would also increase development resources (time & effort) to focus on other things: custom lists (next on the todo list), iOS/iPadOS UI, another refactor/audit of the code, re-writing of SafariConverterLib (BlockKit), re-writing of scriptlet & extendedcss library (if needed, will take a lot of time).

**Comment by stevenya97** [comment]
One area to note is that, while Sequoia has the new feature, it likely won't be back-ported to older macOS versions. Is WebShield going to only be available for Sequoia? Unless we get high Sequoia adoption numbers out the gate, a lot of people will be using Sonoma and older. I think WebShield should still offer a standard zapper solution as fallback. Has the Sequoia feature been rigorously tested with and aginst uBlock's zapper? Until then we can't even be sure most edge cases are covered/blocked and can be considered redundant. Plus developing the zapper tool can help one better understand the quirks of content blocking in Safari in general.

**Reply by arjpar** [reply]
> Is WebShield going to only be available for Sequoia?

@stevenya97 I plan to only support the latest Safari & macOS, but given that the content blocking API hasn't change for a few years, I believe the earliest I could support would be Big Sur. The only remaining issue for Big Sur support would be how much JS Big Sur-era WebKit supports. Given that I plan to transition the scriptlet & extendedcss library to TypeScript, supporting an older era of WebKit shouldn't be too hard of a problem.

One of the reasons why I want(ed) to support latest macOS/iOS is because I want to use the latest content blocking, Safari extension & js features to provide the best experience. But it seems like they haven't been changing too much for the past few years, so I can certainly support iOS/macOS/etc that was released around ~2020 it seems like.

> Has the Sequoia feature been rigorously tested with and aginst uBlock's zapper?

It's still in beta, so testing is being done & feedback is being received. The final product will be polished. Apple usually starts out with a rudimentary feature, and then improves it greatly over time. Given that I haven't seen anyone who hates or dislikes the Safari-native "element zapper", I doubt they would remove it, especially since it ties into their value of privacy.

**Comment by jeremybahadirli** [comment]
I support removal of the Element Zapper feature.

**Comment by stevenya97** [comment]
Safari 18 release is now available for Sequoia, Sonoma, and Ventura. [Link to Ars Technica article](https://arstechnica.com/gadgets/2024/09/macos-15-sequoia-the-ars-technica-review/10/#h1). Running it on my personal machine and seems to work great so far. This is great as a lot more people potentially have access to the Safari zapper feature than previously discussed. I think the discussion can be made again to remove the Element Zapper feature in favor of directing users to use Safari zapper instead. It should be noted that for dynamic ad content that doesn't rely on a static element, such ads may circumvent the zapper and reappear with page reloads.

**Reply by arjpar** [reply]
sounds good. given i dont want to recreate a feature that is already provided in Safari, Ill probably not implement the element zapper. this will give me time to do the other things on the todo list (custom filter lists is what im working on now)

**Comment by arjpar** [comment]
I have decided not to recreate the Element Zapper, as Safari already provides a native one. I doubt Apple will remove a feature that many users appreciate, especially given their emphasis on privacy. Moreover, I expect Apple's Element Zapper to improve over time, potentially leveraging WebKit capabilities unavailable through public APIs. In other words, Safari’s native Element Zapper may be a superior implementation compared with those from others like AdGuard or 1Blocker. From my experience with adblockers in Safari, including AdGuard and 1Password, Safari’s implementation of an Element Zapper is the best.
